### PR TITLE
Extend E0623 for fn items

### PR DESCRIPTION
**Pull request by gaurikholkar-zz**
This fixes #44516 
The below example now gives
```
error[E0623]: lifetime mismatch
 --> gg.rs:3:10
  |
2 | fn foo(x:fn(&u8, &u8), y: Vec<&u8>, z: &u8) {
  |                               ---      --- these two types are declared with different lifetimes...
3 |   y.push(z);
  |          ^ ...but data from `z` flows into `y` here

error: aborting due to previous error
```

r? @nikomatsakis 
cc @arielb1

### CODE REVIEW

File: src/librustc/infer/error_reporting/different_lifetimes.rs [hunk 1]
```diff
@@ -204,6 +204,13 @@ impl<'a, 'gcx, 'tcx> Visitor<'gcx> for FindNestedTypeVisitor<'a, 'gcx, 'tcx> {
 
     fn visit_ty(&mut self, arg: &'gcx hir::Ty) {
         match arg.node {
+            hir::TyBareFn(_) => {
+                self.depth += 1;
+                intravisit::walk_ty(self, arg);
+                self.depth -= 1;
+                return;
+            }
+
```
Comment (kennytm) [inline src/librustc/infer/error_reporting/different_lifetimes.rs]: Trailing whitespace

```
[00:04:19] tidy error: /checkout/src/librustc/infer/error_reporting/different_lifetimes.rs:213: trailing whitespace
```

BTW you could run the tidy test locally using 

```sh
./x.py test src/tools/tidy
```

File: src/test/ui/lifetime-errors/ex3-both-anon-regions-one-is-struct-3.rs [hunk 2]
```diff
@@ -11,7 +11,11 @@
 struct Ref<'a, 'b> { a: &'a u32, b: &'b u32 }
 
 fn foo(mut y: Ref, x: &u32) {
+<<<<<<< HEAD
     y.b = x;
+=======
+    x = y.b;
+>>>>>>> Adding E0623 for structs
```
Comment (kennytm) [inline src/test/ui/lifetime-errors/ex3-both-anon-regions-one-is-struct-3.rs]: ????

**Review by kennytm — CHANGES_REQUESTED** [review]

File: src/librustc/infer/error_reporting/different_lifetimes.rs (hunk 1 above)
Comment (arielb1) [reply src/librustc/infer/error_reporting/different_lifetimes.rs]: You also need to handle `TyTraitObject` (aka the `Fn` in `Box<Fn<(&u8, &u8)>>`).

File: src/test/ui/lifetime-errors/ex3-both-anon-regions-one-is-struct-3.rs (hunk 2 above)
Comment (arielb1) [reply src/test/ui/lifetime-errors/ex3-both-anon-regions-one-is-struct-3.rs]: ERROR: merge conflict marker in file

File: src/librustc/infer/error_reporting/different_lifetimes.rs (hunk 1 above)
Comment (gaurikholkar) [reply src/librustc/infer/error_reporting/different_lifetimes.rs]: yeah at it :)

Comment (gaurikholkar) [reply src/librustc/infer/error_reporting/different_lifetimes.rs]: @arielb1 need an example to test

File: src/librustc/infer/error_reporting/different_lifetimes.rs [hunk 3]
```diff
@@ -204,6 +204,13 @@ impl<'a, 'gcx, 'tcx> Visitor<'gcx> for FindNestedTypeVisitor<'a, 'gcx, 'tcx> {
 
     fn visit_ty(&mut self, arg: &'gcx hir::Ty) {
         match arg.node {
+            hir::TyBareFn(_) => {
+                self.depth += 1;
+                intravisit::walk_ty(self, arg);
+                self.depth -= 1;
```
Comment (kennytm) [inline src/librustc/infer/error_reporting/different_lifetimes.rs]: No such field `depth`. Do you mean the depth of the de Bruijn index or something?

```rust
[00:06:31] error[E0609]: no field `depth` on type `&mut infer::error_reporting::different_lifetimes::FindNestedTypeVisitor<'a, 'gcx, 'tcx>`
[00:06:31]    --> /checkout/src/librustc/infer/error_reporting/different_lifetimes.rs:208:22
[00:06:31]     |
[00:06:31] 208 |                 self.depth += 1;
[00:06:31]     |                      ^^^^^
[00:06:31] 
[00:06:31] error[E0609]: no field `depth` on type `&mut infer::error_reporting::different_lifetimes::FindNestedTypeVisitor<'a, 'gcx, 'tcx>`
[00:06:31]    --> /checkout/src/librustc/infer/error_reporting/different_lifetimes.rs:210:22
[00:06:31]     |
[00:06:31] 210 |                 self.depth -= 1;
[00:06:31]     |                      ^^^^^
[00:06:31] 
[00:06:41] error: aborting due to 2 previous errors
[00:06:41] 
[00:06:41] error: Could not compile `rustc`.
```

**Review by kennytm — CHANGES_REQUESTED** [review]

File: src/test/ui/lifetime-errors/ex3-both-anon-regions-both-are-structs-4.stderr [hunk 4]
```diff
@@ -0,0 +1,12 @@
+error[E0623]: lifetime mismatch
+  --> $DIR/ex3-both-anon-regions-both-are-structs-4.rs:16:11
+   |
+15 | fn foo(mut x: Ref) {
+   |               ---
+   |               |
+   |               these two types are declared with different lifetimes...
```
Comment (kennytm) [inline src/test/ui/lifetime-errors/ex3-both-anon-regions-both-are-structs-4.stderr]: Test case differs from actual output.

```diff
 error[E0623]: lifetime mismatch
   --> $DIR/ex3-both-anon-regions-both-are-structs-4.rs:16:11
    |
 15 | fn foo(mut x: Ref) {
    |               ---
    |               |
-   |               these two types are declared with different lifetimes...
+   |               this type was declared with multiple lifetimes...
 16 |     x.a = x.b;
    |           ^^^ ...but data with one lifetime flows into the other here
```

File: src/test/ui/lifetime-errors/ex3-both-anon-regions-using-fn-items.stderr [hunk 5]
```diff
@@ -0,0 +1,10 @@
+error[E0623]: lifetime mismatch
+  --> src/test/ui/lifetime-errors/ex3-both-anon-regions-using-fn-items.rs:11:10
```
Comment (kennytm) [inline src/test/ui/lifetime-errors/ex3-both-anon-regions-using-fn-items.stderr]: Ditto

```diff
 error[E0623]: lifetime mismatch
-  --> src/test/ui/lifetime-errors/ex3-both-anon-regions-using-fn-items.rs:11:10
+  --> $DIR/ex3-both-anon-regions-using-fn-items.rs:11:10
    |
 10 | fn foo(x:fn(&u8, &u8), y: Vec<&u8>, z: &u8) {
    |                               ---      --- these two types are declared with different lifetimes...
 11 |   y.push(z);
    |          ^ ...but data from `z` flows into `y` here
 
 error: aborting due to previous error
```

**Review by kennytm — CHANGES_REQUESTED** [review]

**Comment by arielb1** [comment]
I think this doesn't handle the case of structs
```Rust
struct Ref<'a>(&'a ());
fn foo(x:fn(Ref, Ref), y: Vec<&u8>, z: &u8) {
  y.push(z);
}
```

**Comment by arielb1** [comment]
@bors p=1

This is blocking further work

**Comment by gaurikholkar** [comment]
@arielb1 planning to handle it in a separate PR.

File: fn.rs [hunk 6]
```diff
@@ -0,0 +1,8 @@
+
+fn foo(x:  fn(&u8, &u8), y: Vec<&u8>, z: &u8) {
```
Comment (bjorn3) [inline fn.rs]: Did you accidentially left this file in the root dir?

Comment (gaurikholkar) [reply fn.rs]: Oh yeah. I'm open a PR for that or I have an ongoing PR, I'll make sure to delete that.

Comment (gaurikholkar) [reply fn.rs]: Fixed in #44549. Thanks!

**Pull request merged**